### ISSUE TITLE
Latest hangs during linking

### ISSUE THREAD

**Issue by Photosounder** [emscripten-core/emsdk#821]
When I build my project with the latest version of emcc it hangs at a linking stage indefinitely. When I revert to 2.0.20 and build the same project the same way everything works fine.

**Comment by sbc100** [comment]
Can you run with `-v` or `EMCC_DEBUG=1` do find out exactly which part of the linker is hanging?  

Also, if you could provide your link flags that might give us a clue?

**Comment by Photosounder** [comment]
My whole building command is this:
`EMCC_DEBUG=1 emcc -v rl.c commands.c drawobjects.c game.c geometry.c graphics.c gui.c main.c opencl.c science.c sinogram.c sound.c util.c -Wno-dangling-else -Wno-pointer-sign -Wno-incompatible-pointer-types -Wno-parentheses -ferror-limit=1 --embed-file ../vector_type -s USE_SDL=2 -s ENVIRONMENT=web,worker -s ALLOW_MEMORY_GROWTH=0 -s INITIAL_MEMORY=268435456 -s USE_PTHREADS=1 -s PTHREAD_POOL_SIZE=4 -O3 -g0 --source-map-base https://photosounder.com/emscripten/src/spacewar/ -o spacewar.html`

When it hangs it says this:
`emcc:DEBUG: linking: ['/tmp/emscripten_temp/rl_0.o', '/tmp/emscripten_temp/commands_1.o', '/tmp/emscripten_temp/drawobjects_2.o', '/tmp/emscripten_temp/game_3.o', '/tmp/emscripten_temp/geometry_4.o', '/tmp/emscripten_temp/graphics_5.o', '/tmp/emscripten_temp/gui_6.o', '/tmp/emscripten_temp/main_7.o', '/tmp/emscripten_temp/opencl_8.o', '/tmp/emscripten_temp/science_9.o', '/tmp/emscripten_temp/sinogram_10.o', '/tmp/emscripten_temp/sound_11.o', '/tmp/emscripten_temp/util_12.o', '-L/home/user/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten', '/home/user/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libSDL2-mt.a', '/home/user/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/crtbegin.o', '/home/user/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libgl-mt.a', '/home/user/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libal.a', '/home/user/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libhtml5.a', '/home/user/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-mt.a', '/home/user/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libcompiler_rt-mt.a', '/home/user/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc++-mt-noexcept.a', '/home/user/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc++abi-mt-noexcept.a', '/home/user/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libdlmalloc-mt.a', '/home/user/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc_rt_wasm.a', '/home/user/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libsockets-mt.a']
 "/home/user/emsdk/upstream/bin/wasm-ld" -o spacewar.wasm /tmp/emscripten_temp/rl_0.o /tmp/emscripten_temp/commands_1.o /tmp/emscripten_temp/drawobjects_2.o /tmp/emscripten_temp/game_3.o /tmp/emscripten_temp/geometry_4.o /tmp/emscripten_temp/graphics_5.o /tmp/emscripten_temp/gui_6.o /tmp/emscripten_temp/main_7.o /tmp/emscripten_temp/opencl_8.o /tmp/emscripten_temp/science_9.o /tmp/emscripten_temp/sinogram_10.o /tmp/emscripten_temp/sound_11.o /tmp/emscripten_temp/util_12.o -L/home/user/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten /home/user/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libSDL2-mt.a /home/user/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/crtbegin.o /home/user/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libgl-mt.a /home/user/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libal.a /home/user/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libhtml5.a /home/user/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-mt.a /home/user/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libcompiler_rt-mt.a /home/user/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc++-mt-noexcept.a /home/user/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc++abi-mt-noexcept.a /home/user/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libdlmalloc-mt.a /home/user/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc_rt_wasm.a /home/user/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libsockets-mt.a -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --allow-undefined --import-memory --shared-memory --strip-debug --export main --export stackSave --export stackRestore --export stackAlloc --export __wasm_call_ctors --export __errno_location --export __emscripten_pthread_data_constructor --export __pthread_tsd_run_dtors --export _emscripten_call_on_thread --export _emscripten_do_dispatch_to_thread --export _emscripten_main_thread_futex --export _emscripten_thread_init --export emscripten_current_thread_process_queued_calls --export _emscripten_allow_main_runtime_queued_calls --export emscripten_futex_wake --export emscripten_get_global_libc --export emscripten_main_browser_thread_id --export emscripten_main_thread_process_queued_calls --export emscripten_register_main_browser_thread_id --export emscripten_run_in_main_runtime_thread_js --export emscripten_stack_set_limits --export emscripten_sync_run_in_main_thread_2 --export emscripten_sync_run_in_main_thread_4 --export emscripten_tls_init --export pthread_self --export memalign --export malloc --export free --export memset --export memcpy --export _get_tzname --export _get_daylight --export _get_timezone --export-if-defined=__start_em_asm --export-if-defined=__stop_em_asm --export-table -z stack-size=5242880 --initial-memory=268435456 --no-entry --max-memory=268435456 --global-base=1024`

When I send Ctrl-C it says this:
`^Cprofiler:DEBUG: block "link" took 171.77 seconds
emcc:WARNING: KeyboardInterrupt`

**Comment by Photosounder** [comment]
I'm not sure how I'm supposed to do this. I tried:
`EMCC_DEBUG=1 emcc -v rl.c commands.c drawobjects.c game.c geometry.c graphics.c gui.c main.c opencl.c science.c sinogram.c sound.c util.c -Wno-dangling-else -Wno-pointer-sign -Wno-incompatible-pointer-types -Wno-parentheses -ferror-limit=1 --embed-file ../vector_type -s USE_SDL=2 -s ENVIRONMENT=web,worker -s ALLOW_MEMORY_GROWTH=0 -s INITIAL_MEMORY=268435456 -s USE_PTHREADS=1 -s PTHREAD_POOL_SIZE=4 -O0 -Wl,-O0 -g0 --source-map-base https://photosounder.com/emscripten/src/spacewar/ -o spacewar.html`
but it doesn't change anything.

**Comment by sbc100** [comment]
If you pass `-Wl,-O0` and the problem persists then the issue is probably not related to string merging after all.

Is there any chance you could share a repro case?

**Comment by Photosounder** [comment]
I'm trying to make one based on a simpler but similar project but I keep getting errors like
```
[parse exception: empty function size (at 0:919707)]
Fatal: error in parsing input
emcc: error: '/home/user/emsdk/upstream/bin/wasm-emscripten-finalize --minimize-wasm-changes --dyncalls-i64 minimal_viewer.wasm -o minimal_viewer.wasm --detect-features' failed (1)
```
even with version 2.0.20 and I have no idea how to deal with this error.

**Comment by sbc100** [comment]
That looks like a separate issue that we should fix first .. wasm-emscripten-finalize should never fail in that way.   Do you have a repro case for that error?

**Comment by Photosounder** [comment]
Yes I do. Hopefully it's reproducible right out of the box just by extracting the zip and running `./embuilsh.sh`.
[emcc 2.0.21 bug.zip](https://github.com/emscripten-core/emsdk/files/6510570/emcc.2.0.21.bug.zip)

**Comment by sbc100** [comment]
Interesting .. it seems to work fine for me (with `emsdk install 2.0.21`):

```
$ time sh -x embuild.sh 
+ emcc rl.c minimal_viewer.c -Wno-dangling-else -Wno-pointer-sign -Wno-incompatible-pointer-types -Wno-parentheses -ferror-limit=1 -s USE_SDL=2 -s ENVIRONMENT=web,worker -s ALLOW_MEMORY_GROWTH=0 -s INITIAL_MEMORY=268435456 -s USE_PTHREADS=1 -s PTHREAD_POOL_SIZE=4 -O2 -g0 -o minimal_viewer.html
In file included from rl.c:2:
In file included from ./rouziclib/rouziclib.c:149:
In file included from ./rouziclib/libraries/zlib.c:3:
./rouziclib/libraries/orig/miniz.c:3031:9: warning: Using fopen, ftello, fseeko, stat() etc. path for file I/O - this path may not support large files. [-W#pragma-messages]
#pragma message("Using fopen, ftello, fseeko, stat() etc. path for file I/O - this path may not support large files.")
        ^
'+sse4.1' is not a recognized feature for this target (ignoring feature)
'+sse4.1' is not a recognized feature for this target (ignoring feature)
'+ssse3' is not a recognized feature for this target (ignoring feature)
'+ssse3' is not a recognized feature for this target (ignoring feature)
1 warning generated.

real	0m23.991s
user	0m47.984s
sys	0m0.780s
```

(I had to remove the --embed-file part from the command line but I don't think that is relevant)

**Comment by Photosounder** [comment]
I checked (and installed 2.0.21 specifically) and for me the problem persists.

**Comment by sbc100** [comment]
And you are running the exact same shell script, and on linux too?

**Comment by kripken** [comment]
It should be practical to disagnose a hang directly. The `EMCC_DEBUG=1` output above has a wasm-ld command (the long line that starts with `"/home/user/emsdk/upstream/bin/wasm-ld" -o spacewar.wasm`). You can run that in a debugger like `gdb`, then after you are sure it is hanging you can pause it and get a stack trace. Then we would know exactly what it is doing when it hangs.

**Comment by Photosounder** [comment]
> And you are running the exact same shell script, and on linux too?

I think it might be an I/O problem. I'm running Mint in a VirtualBox VM and usually directly compiling from a mounted Windows folder (because I do all the programming in Windows). When I download the ZIP in Windows and try to compile it directly in Mint from the mounted folder I get the same problem, however when I download the ZIP in Mint, extract it to the Mint desktop and compile it there it compiles correctly. Doing it in a mounted folder seems like something that shouldn't be a problem (maybe it messes with POSIX file mapping?), and yet it is.

**Comment by sbc100** [comment]
We have heard before of issues with VirtualBox VM running on windows.   IIRC they only occur when the wasm-ld attempts to write directly the a windows drive mount.    

It is odd that the loop loop should be in `llvm::StringTableBuilder` since that is something we just started using in 2.0.21.    So its odd to me that `-Wl,-O0` (which disables the StringTableBuilder) doesn't happen to fix the problem.

A couple more questions:

1. Is thread5 blocked on the memcpy or is it spinning/busy?
2. What does the backtrace of thread 5 (the non-terminating thread) look like when you add `-Wl,-O0`?
3. How does passing `-Wl,--threads=1` effect the problem?  (this removes the multi-threading from wasm-ld).

**Comment by Photosounder** [comment]
I see, I started using a VM because a few years ago I had issues with compilation not really working on native Windows. I guess I'll have to mirror files.

1. Thread 5 isn't blocked, it makes constant memcpy calls, I can step through them.
2. I guess I didn't wait long enough when I first tried this (because it takes a few minutes at the same stage so I thought that was it). I did that and now it spinned for a long time in `lld::wasm::InputFunction::writeTo` or `lld::wasm::InputChunk::writeTo` also doing memcpys while blocking every other thread but this time it completed after a few minutes. So it works but takes a while (total compilation is 6:07 for `latest`, 7:10 for `2.0.20` and 6:37 without `-Wl,-O0` in `2.0.20`).
3. `-Wl,--threads=1` without `-Wl,-O0` completes the `wasm-ld` command, after 51 minutes! I was about to give up and kill it when it ended. It says `profiler:DEBUG: block "link" took 2992.41 seconds`. Then it failed with `[parse exception: bad user section size (at 0:12231907)]`, but that's with me still using the mounted Windows folder, although oddly that wasn't a problem when compiling the same project in the same way as I did in paragraph `2.`.

I'll try running the original build command for a few hours to see if it ever completes.

**Comment by Photosounder** [comment]
I let it run with the original command and the compilation does complete, albeit also in 51 minutes (it says `profiler:DEBUG: block "link" took 2927.19 seconds), which is odd and some kind of a regression. The exact same command takes only 7 minutes with `2.0.20` (it says `profiler:DEBUG: block "link" took 254.86 seconds`).

When I change the output target to `-o ~/emcc_output/spacewar.html` with `2.0.20` it says `profiler:DEBUG: block "link" took 0.44 seconds` (and the whole command runs in 2½ minutes instead of 7) and with `latest` it says `profiler:DEBUG: block "link" took 0.13 seconds` and the whole command runs in 2 minutes and 21 seconds instead of 51 minutes.

As for the other project which failed compilation entirely, now when I output to `-o ~/emcc_output/minimal_viewer.html` the whole compilation completes correctly in 37 seconds. So it seems like outputting to a mounted Windows folder is the root of all the problems I described and should always be avoided. I don't know if there's a way you can test if the output target is suitable or not, but if possible a warning would have been very welcome, although now that I know what the problem is I for one don't need such a warning.

**Comment by sbc100** [comment]
If I had to guess this is an issue with the way VirtualBox works with mmap'd files from windows filesystems.   The linker writes to this memory map in parallel is is likely to hammer it pretty hard.   It seems like its not just slow but also causing real errors :( 

I found some other old issues that mention VirtualBox:

https://github.com/emscripten-core/emscripten/issues/4480
https://github.com/emscripten-core/emsdk/issues/677
